File: app/src/main/java/com/cosage/zzh/wxmusic/WeChatMessage.java
package com.cosage.zzh.wxmusic;

import android.content.ContentValues;
import android.support.test.uiautomator.UiObjectNotFoundException;

import com.gh0u1l5.wechatmagician.spellbook.base.Operation;
import com.gh0u1l5.wechatmagician.spellbook.interfaces.IDatabaseHook;

import org.jetbrains.annotations.NotNull;

import de.robv.android.xposed.XposedBridge;

import static de.robv.android.xposed.XposedBridge.log;

/**
 * Created by Zhengzhihui on 2018/5/28.
 */

public class WeChatMessage implements IDatabaseHook {


    public static WeChatMessage weChatMessage = null;
    //public boolean isPlayMusic=false;
    public static WeChatMessage instance() {
        if (weChatMessage == null) {
            weChatMessage = new WeChatMessage();
        }
        return weChatMessage;
    }

    public static final String pageName = "com.tencent.mm";

    @NotNull
    @Override
    public Operation<Object> onDatabaseOpening(String s, Object o, int i, Object o1) {
        return null;
    }

    @NotNull
    @Override
    public Operation<Object> onDatabaseOpened(String s, Object o, int i, Object o1, Object o2) {
        return null;
    }

    @NotNull
    @Override
    public Operation<Object> onDatabaseQuerying(Object o, Object o1, String s, String[] strings, String s1, Object o2) {
        return null;
    }

    @NotNull
    @Override
    public Operation<Object> onDatabaseQueried(Object o, Object o1, String s, String[] strings, String s1, Object o2, Object o3) {
        return null;
    }

    @NotNull
    @Override
    public Operation<Long> onDatabaseInserting(Object o, String s, String s1, ContentValues contentValues, int i) {
        return null;
    }

    @NotNull
    @Override
    public Operation<Long> onDatabaseInserted(Object o, String table, String s1, ContentValues contentValues, int i, Long aLong) {
         if (table.equals( "message") && contentValues.getAsInteger("isSend") == 0) {
             String content= contentValues.getAsString("content");
             if(content.contains("发给我")) {
                 try {
                     ShellUtils.execCommand("am instrument -w -r   -e debug false -e class com.cosage.zzh.uitest.WxTest com.cosage.zzh.uitest.test/android.support.test.runner.AndroidJUnitRunner", true);
                     ;
                 } catch (Exception e) {
                     XposedBridge.log("error:" + e.getMessage());
                 }
             }
            /*String content = contentValues.getAsString("content");
            if(content.contains("#点歌") && !isPlayMusic){
                String[] strs = content.split("#点歌");
                if(strs.length>=2){
                    isPlayMusic = true;
                    yunMusicUI.playMusic(strs[1]);
                }
            }*/
            log("New Message: " + contentValues.getAsString("content"));
        }
        return null;
    }


    @NotNull
    @Override
    public Operation<Integer> onDatabaseUpdating(Object o, String s, ContentValues contentValues, String s1, String[] strings, int i) {
        return null;
    }

    @NotNull
    @Override
    public Operation<Integer> onDatabaseUpdated(Object o, String s, ContentValues contentValues, String s1, String[] strings, int i, int i1) {
        return null;
    }

    @NotNull
    @Override
    public Operation<Integer> onDatabaseDeleting(Object o, String s, String s1, String[] strings) {
        return null;
    }

    @NotNull
    @Override
    public Operation<Integer> onDatabaseDeleted(Object o, String s, String s1, String[] strings, int i) {
        return null;
    }

    @Override
    public boolean onDatabaseExecuting(Object o, String s, Object[] objects, Object o1) {
        return false;
    }

    @Override
    public void onDatabaseExecuted(Object o, String s, Object[] objects, Object o1) {

    }
}

File: app/src/main/java/com/cosage/zzh/wxmusic/WeChatMessage.kt
package com.cosage.zzh.wxmusic

import android.content.ContentValues
import android.support.test.uiautomator.UiObjectNotFoundException
import android.widget.BaseAdapter

import com.gh0u1l5.wechatmagician.spellbook.base.Operation
import com.gh0u1l5.wechatmagician.spellbook.hookers.ListViewHider
import com.gh0u1l5.wechatmagician.spellbook.interfaces.IAdapterHook
import com.gh0u1l5.wechatmagician.spellbook.interfaces.IDatabaseHook

import java.io.IOException

import de.robv.android.xposed.XposedBridge

import de.robv.android.xposed.XposedBridge.log
import de.robv.android.xposed.XposedHelpers.getObjectField

/**
 * Created by Zhengzhihui on 2018/5/28.
 */

object WeChatMessage : IDatabaseHook,IAdapterHook {

    @Volatile
    var isHanding = false

    val pageName = "com.tencent.mm"

    override fun onDatabaseInserted(thisObject: Any, table: String, nullColumnHack: String?, initialValues: ContentValues?, conflictAlgorithm: Int, result: Long?): Operation<Long> {
        if (table != "message") {
            return super.onDatabaseInserted(thisObject, table, nullColumnHack, initialValues, conflictAlgorithm, result)
        }
        if (initialValues?.getAsInteger("isSend") != -1) {
            XposedBridge.log("新消息:" + initialValues.toString())
            //测试

            if (isHanding) {
                if (initialValues?.getAsInteger("isSend") == 1) {
                    isHanding = false;
                }
                if (isHanding) {
                    return super.onDatabaseInserted(thisObject, table, nullColumnHack, initialValues, conflictAlgorithm, result)
                }
            }
            var content = initialValues?.getAsString("content");
            //if (content.contains("#切歌")) {
            if (content?.contains("#切歌") == true) {
                try {
                    XposedBridge.log("开始切歌");
                    isHanding = true;
                    val musicName = content.substring(content.indexOf("#切歌") + 3);
                    //String musicName= "童话镇3";
                    XposedBridge.log("musicName:" + musicName);
                    Http.put("name", musicName);
                    Http.put("reply", "切歌成功：正在播放~~" + musicName);
                    ShellUtils.execCommand("am instrument -w -r   -e debug false -e class com.cosage.zzh.wxmusic.ExampleInstrumentedTest#useAppContext com.cosage.zzh.wxmusic.test/android.support.test.runner.AndroidJUnitRunner", true);
                } catch (e: Exception) {

                }
            } else if (content?.contains("#购买") == true) {
                XposedBridge.log("购买搜索");
                isHanding = true;
                var userName = content.substring(content.indexOf("#购买") + 3);


            }
            log("New Message: " + initialValues?.getAsString("content"));
        } else {
            isHanding = false
        }
        return super.onDatabaseInserted(thisObject, table, nullColumnHack, initialValues, conflictAlgorithm, result)
    }

    override fun onConversationAdapterCreated(adapter: BaseAdapter) {

        ListViewHider.register(adapter, "Chatroom Hider") { item ->
            val username = getObjectField(item, "field_username")
            XposedBridge.log("username == " + username)
            false
            //username in ChatroomHideList
        }
    }


    override fun onDatabaseQueried(thisObject: Any, factory: Any?, sql: String, selectionArgs: Array<String>?, editTable: String?, cancellationSignal: Any?, result: Any?): Operation<Any> {
        XposedBridge.log("q:"+editTable)
        return super.onDatabaseQueried(thisObject, factory, sql, selectionArgs, editTable, cancellationSignal, result)
    }
}
